challenge-two/src/agent.ts

```diff
@@ -5,85 +5,85 @@ import {
+// Load environment variables from .env file
+dotenv.config();
+
+const { RPC_PROVIDER_URL } = process.env;
 
 export const provideHandleTransaction = (
-  abi: any,
-  swapRouter02Address: string
+  provider: ethers.providers.Provider | any
 ): HandleTransaction => {
   return async (txEvent: TransactionEvent) => {
     const findings: Finding[] = [];
-    const lowerCaseSwapRouter02Address = swapRouter02Address.toLowerCase();
-
-    // Filter transaction events by logs that match the swap events and router address
-    const swapLogs = txEvent.filterFunction(abi, lowerCaseSwapRouter02Address);
 
-    swapLogs.forEach((swap) => {
-      if (swap?.name === "exactInputSingle") {
-        // Extract swap event arguments
-        const tokenIn = swap.args.params["tokenIn"];
-        const tokenOut = swap.args.params["tokenOut"];
-        const fee = swap.args.params["fee"];
-        const recipient = swap.args.params["recipient"];
-        const amountIn = swap.args.params["amountIn"];
-        const amountOutMinimum = swap.args.params["amountOutMinimum"];
-        const sqrtPriceLimitX96 = swap.args.params["sqrtPriceLimitX96"];
+    // Filter logs for Swap events by checking the first topic (event signature)
+    const swapLogs = txEvent.logs.filter(
+      (log) => log.topics[0] === SWAP_EVENT_SIGNATURE
+    );
 
-        findings.push(
-          Finding.fromObject({
-            alertId: "FORTA-2",
-            name: "Nethermind Forta Bot UniSwap - exactInputSingle",
-            description: `Swap event detected on ${swapRouter02Address}`,
-            severity: FindingSeverity.Low,
-            type: FindingType.Info,
-            metadata: {
-              tokenIn: tokenIn,
-              tokenOut: tokenOut,
-              fee: fee.toString(),
-              recipient,
-              amountIn: amountIn.toString(),
-              amountOutMinimum: amountOutMinimum.toString(),
-              sqrtPriceLimitX96: sqrtPriceLimitX96.toString(),
-            },
-          })
+    for (const log of swapLogs) {
+      try {
+        const poolContract = new ethers.Contract(
+          log.address,
+          UNISWAP_V3_POOL_ABI,
+          provider
         );
-      } else if (swap?.name === "exactOutputSingle") {
-        const tokenIn = swap.args.params["tokenIn"];
-        const tokenOut = swap.args.params["tokenOut"];
-        const fee = swap.args.params["fee"];
-        const recipient = swap.args.params["recipient"];
-        const amountOut = swap.args.params["amountOut"];
-        const amountInMaximum = swap.args.params["amountInMaximum"];
-        const sqrtPriceLimitX96 = swap.args.params["sqrtPriceLimitX96"];
+
+        // Check if the log is from a valid Uniswap V3 pool
+        const poolFactory: string = await poolContract.callStatic.factory();
+
+        if (
+          poolFactory.toLowerCase() !== UNISWAP_V3_FACTORY_ADDRESS.toLowerCase()
+        ) {
+          // This is not a valid Uniswap V3 pool
+          continue;
+        }
```

Comment:
Here, this check may not work as you expect, since _any_ contract that adheres to the UniswapV3 pool ABI could return the UniV3 factory address even if it wasn't created by it. The bot should only detect `swap`s from pool that were created by the UniV3 factory.